### PR TITLE
status: Add experimental cluster JSON status with state transition support

### LINKED ISSUE #8730: Add transient states ("stopping", "starting")
Minikube status lists Running, Paused or Stopped for each profile.  It does not account for the state between minikube start and Running.  It would be helpful to indicate when minikube is starting, pausing (maybe, this state seems like it might be super short), stopping and possibly deleting as well as valid states.

### PR DESCRIPTION
**Pull request by tstromberg**
This implements state transitions on top of the recent CloudEvents JSON output work. Fixes #8730 

 In a nutshell:

* minikube status now supports a `--layout=cluster`, which supports a status fields along with the current step:

```
{
  "Name": "minikube",
  "StatusCode": 418,
  "StatusName": "Paused",
  "Step": "Done",
  "StepDetail": "⏯️  Paused 24 containers in: kube-system, kubernetes-dashboard, storage-gluster, istio-operator",
  "BinaryVersion": "v1.12.1",
  "Components": {
    "kubeconfig": {
      "Name": "kubeconfig",
      "StatusCode": 200,
      "StatusName": ""
    }
  },
  "Nodes": [
    {
      "Name": "minikube",
      "StatusCode": 200,
      "StatusName": "OK",
      "Components": {
        "apiserver": {
          "Name": "apiserver",
          "StatusCode": 418,
          "StatusName": "Paused"
        },
        "kubelet": {
          "Name": "kubelet",
          "StatusCode": 405,
          "StatusName": "Stopped"
        }
      }
    },
    {
      "Name": "minikube-m02",
      "StatusCode": 200,
      "StatusName": "OK",
      "Components": {
        "kubelet": {
          "Name": "kubelet",
          "StatusCode": 405,
          "StatusName": "Stopped"
        }
      }
    },
    {
      "Name": "minikube-m03",
      "StatusCode": 200,
      "StatusName": "OK",
      "Components": {
        "kubelet": {
          "Name": "kubelet",
          "StatusCode": 405,
          "StatusName": "Stopped"
        }
      }
    }
  ]
}
```

* CloudEvent steps from the most recent mutation command for a profile are stored in: `$MINIKUBE_HOME/<profile>/events.json`
* Steps are now properly written out for non-start mutations (pause, delete, stop, unpause)
* These events are read back if `--layout=cluster` is invoked.
```

### CODE REVIEW

File: cmd/minikube/cmd/delete.go [hunk 1]
```diff
@@ -44,6 +44,7 @@ import (
 	"k8s.io/minikube/pkg/minikube/localpath"
 	"k8s.io/minikube/pkg/minikube/machine"
 	"k8s.io/minikube/pkg/minikube/out"
+	"k8s.io/minikube/pkg/minikube/out/register"
```
Comment (medyagh) [inline cmd/minikube/cmd/delete.go:47]: irrelelvent, but I wonder if there is a better name or better place for this package ?

something like status or step ... not sure

File: cmd/minikube/cmd/status.go [hunk 2]
```diff
@@ -69,6 +88,36 @@ type Status struct {
 	Worker     bool
 }
 
+// ClusterState holds a cluster state representation
+type ClusterState struct {
```
Comment (medyagh) [inline cmd/minikube/cmd/status.go:92]: ideally I like to see a package called, profile (cluster) that has this in it. but I am okay for it be in CMD with the goal of keeping this PR small

File: cmd/minikube/cmd/status.go [hunk 3]
```diff
@@ -304,3 +362,141 @@ func statusJSON(st []*Status, w io.Writer) error {
 	_, err = w.Write(js)
 	return err
 }
+
+func readEventLog(name string) ([]cloudevents.Event, time.Time, error) {
```
Comment (medyagh) [inline cmd/minikube/cmd/status.go:366]: add comment for this func maybe with an example what it does very short

File: cmd/minikube/cmd/status.go [hunk 4]
```diff
@@ -304,3 +362,141 @@ func statusJSON(st []*Status, w io.Writer) error {
...
+		if err = json.Unmarshal(scanner.Bytes(), &ev); err != nil {
+			return events, st.ModTime(), err
+		}
+		events = append(events, ev)
+	}
+
+	return events, st.ModTime(), nil
+}
+
+func clusterState(sts []*Status) ClusterState {
```
Comment (medyagh) [inline cmd/minikube/cmd/status.go:392]: this one could be moved to a better package in a follow up refactor ? I am ok with it being in cmd now

File: pkg/minikube/out/register/cloud_events.go [hunk 5]
```diff
@@ -44,11 +69,56 @@ func printAsCloudEvent(log Log, data map[string]string) {
...
 	if err != nil {
-		glog.Warningf("error marashalling event: %v", err)
+		glog.Errorf("error marashalling event: %v", err)
+		return
 	}
-	fmt.Fprintln(OutputFile, string(json))
+	fmt.Fprintln(outputFile, string(bs))
+}
+
+func printAndRecordCloudEvent(log Log, data map[string]string) {
```
Comment (medyagh) [inline pkg/minikube/out/register/cloud_events.go:86]: comment here

File: cmd/minikube/cmd/status.go (hunk 4 above)
Comment (tstromberg) [reply cmd/minikube/cmd/status.go:392]: Agreed. I initially had this as part of the `cluster` package, and the node status as part of the `node` package, but I brought it back to `status` to keep this PR small. There was an inverted dependency issue :|

File: cmd/minikube/cmd/delete.go (hunk 1 above)
Comment (tstromberg) [reply cmd/minikube/cmd/delete.go:47]: Agreed.

**Pull request merged**